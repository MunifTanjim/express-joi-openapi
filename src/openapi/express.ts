import { Application, Router } from 'express'
import { RegisteredExpressOpenAPIPlugin } from '../plugins/types'
import { HttpMethod } from '../types'
import { OpenAPISpecification } from './index'
import { Layer, Route, StackLayer, StackName } from './types'

const regexPattern = {
  param: /\(\?:\(\[\^\\\/]\+\?\)\)/,
  path: /^\^\\\/(?:(:?[\w\\.-]*(?:\\\/:?[\w\\.-]*)*)|(\(\?:\(\[\^\\\/]\+\?\)\)))\\/,
  rootPath: /^\^\\\/\?\(\?=\\\/\|\$\)$/,
}

const getPathString = (pathRegex: RegExp, keys: Layer['keys']): string => {
  const pathRegexpSource = pathRegex.source

  let regexString = pathRegexpSource

  let regexStringExecArray = regexPattern.path.exec(regexString)

  let paramIndex = 0
  while (regexPattern.param.test(regexString)) {
    const param = `:${keys[paramIndex].name}`

    regexString = regexString.replace(regexPattern.param, param)

    paramIndex++
  }

  if (regexString !== pathRegexpSource) {
    regexStringExecArray = regexPattern.path.exec(regexString)
  }

  const pathString = regexStringExecArray![1].replace(/\\\//g, '/')

  return pathString
}

const processRoute = (
  spec: OpenAPISpecification,
  route: Route,
  basePath = '',
  plugins: RegisteredExpressOpenAPIPlugin[]
): void => {
  const routePaths = Array.isArray(route.path) ? route.path : [route.path]

  routePaths.forEach((routePath) => {
    const path = `${
      basePath ? `${basePath}${routePath === '/' ? '' : routePath}` : routePath
    }`.replace(/:([a-zA-Z0-9_]+)/g, '{$1}')

    const methods = Object.keys(route.methods).filter(
      (method): method is HttpMethod => method !== '_all'
    )

    if (!spec.paths[path]) {
      spec.setPathItem(path, {})
    }

    for (const method of methods) {
      spec.setPathItemOperation(path, method, {
        responses: {
          default: {
            description: '',
          },
        },
      })

      for (const plugin of plugins) {
        const stashedValue = plugin.stash.find(route, method)
        if (stashedValue) {
          plugin.processRoute(plugin.specification, stashedValue, {
            path,
            method,
          })
        }
      }
    }
  })
}

export const processExpressRouters = (
  spec: OpenAPISpecification,
  app: Application | Router,
  basePath = '',
  plugins: RegisteredExpressOpenAPIPlugin[]
): void => {
  const stack: StackLayer[] = app.stack ?? (app as Application)._router?.stack

  if (!stack) {
    spec.setPathItem(basePath, {})

    return
  }

  for (const layer of stack) {
    // terminal route
    if (layer.route) {
      processRoute(spec, layer.route, basePath, plugins)

      continue
    }

    if (
      layer.name !== StackName.ROUTER &&
      layer.name !== StackName.BOUND_DISPATCH &&
      layer.name !== StackName.MOUNTED_APP
    ) {
      continue
    }

    if (regexPattern.path.test(layer.regexp.source)) {
      const parsedPath = getPathString(layer.regexp, layer.keys)

      processExpressRouters(
        spec,
        layer.handle as Router,
        basePath + '/' + parsedPath,
        plugins
      )

      continue
    }

    if (
      !layer.path &&
      layer.regexp &&
      !regexPattern.rootPath.test(layer.regexp.source)
    ) {
      const regexPath = ' RegExp(' + layer.regexp + ') '

      processExpressRouters(
        spec,
        layer.handle as Router,
        basePath + '/' + regexPath,
        plugins
      )

      continue
    }

    processExpressRouters(spec, layer.handle as Router, basePath, plugins)
  }
}
