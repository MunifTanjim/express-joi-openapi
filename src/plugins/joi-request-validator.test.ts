import express, { Application, ErrorRequestHandler, Handler } from 'express'
import { ExpressOpenAPI } from '../index'
import {
  getJoiRequestValidatorPlugin,
  RequestValidationError,
  GetJoiRequestValidatorMiddleware,
} from './joi-request-validator'
import request from 'supertest'
import Joi from '@hapi/joi'

describe('getJoiRequestValidatorPlugin', () => {
  test('returns plugin', () => {
    const plugin = getJoiRequestValidatorPlugin()
    expect(plugin.name).toMatchInlineSnapshot(`"joi-request-validator"`)
    expect(plugin.getMiddleware).toBeDefined()
    expect(plugin.processRoute).toBeDefined()
  })

  describe('joiRequestValidatorPlugin', () => {
    let validate: GetJoiRequestValidatorMiddleware
    let app: Application

    const handler: Handler = (_req, res) => {
      res.status(200).json({
        pong: 42,
      })
    }

    const errorHandler: ErrorRequestHandler = (err, _req, res, _next) => {
      if (err instanceof RequestValidationError) {
        res.status(400).json({
          error: {
            message: err.message,
            segment: err.segment,
            details: err.validationError.details.map(
              ({ message, path, type }) => ({ message, path, type })
            ),
          },
        })

        return
      }

      res.status(500).json({
        error: {
          message: 'Server Error',
        },
      })
    }

    describe('segment validation', () => {
      beforeEach(() => {
        const expressOpenApi = new ExpressOpenAPI()

        app = express()
        app.use(express.json())

        const joiRequestValidatorPlugin = getJoiRequestValidatorPlugin()

        validate = expressOpenApi.registerPlugin(joiRequestValidatorPlugin)
      })

      test('validates headers', async () => {
        app.post(
          '/ping',
          validate({
            headers: Joi.object({
              authorization: Joi.string()
                .pattern(/^Bearer .+$/)
                .required(),
            }).unknown(),
          }),
          handler
        )

        app.use(errorHandler)

        const invalidResponse1 = await request(app)
          .post('/ping')
          .set('authorization', `JWT 42`)
        expect(invalidResponse1.status).toBe(400)
        expect(invalidResponse1.body).toMatchSnapshot()

        const validResponse1 = await request(app)
          .post('/ping')
          .set('authorization', `Bearer 42`)
        expect(validResponse1.status).toBe(200)
      })

      test('validates params', async () => {
        app.post(
          '/ping/:id',
          validate({
            params: Joi.object({
              id: Joi.number().integer().required(),
            }),
          }),
          handler
        )

        app.use(errorHandler)

        const invalidResponse1 = await request(app).post('/ping/forty-two')
        expect(invalidResponse1.status).toBe(400)
        expect(invalidResponse1.body).toMatchSnapshot()

        const validResponse1 = await request(app).post('/ping/42')
        expect(validResponse1.status).toBe(200)
      })

      test('validates query', async () => {
        app.post(
          '/ping',
          validate({
            query: Joi.object({
              count: Joi.number().integer().required(),
            }),
          }),
          handler
        )

        app.use(errorHandler)

        const invalidResponse1 = await request(app).post('/ping')
        expect(invalidResponse1.status).toBe(400)
        expect(invalidResponse1.body).toMatchSnapshot()

        const invalidResponse2 = await request(app).post(
          '/ping?count=forty-two'
        )
        expect(invalidResponse2.status).toBe(400)
        expect(invalidResponse2.body).toMatchSnapshot()

        const validResponse1 = await request(app).post('/ping?count=42')
        expect(validResponse1.status).toBe(200)
      })

      test('validates body', async () => {
        app.post(
          '/ping',
          validate({
            body: {
              type: Joi.string()
                .valid('plastic', 'rubber', 'wood')
                .default('rubber'),
            },
          }),
          handler
        )

        app.use(errorHandler)

        const invalidResponse1 = await request(app)
          .post('/ping')
          .send({ type: 'paper' })
        expect(invalidResponse1.status).toBe(400)
        expect(invalidResponse1.body).toMatchSnapshot()

        const validResponse1 = await request(app).post('/ping')
        expect(validResponse1.status).toBe(200)

        const validResponse2 = await request(app)
          .post('/ping')
          .send({ type: 'wood' })
        expect(validResponse2.status).toBe(200)
      })
    })

    describe('value coercion', () => {
      const mockFn = jest.fn()

      beforeEach(() => {
        const expressOpenApi = new ExpressOpenAPI()

        app = express()
        app.use(express.json())

        const joiRequestValidatorPlugin = getJoiRequestValidatorPlugin()

        validate = expressOpenApi.registerPlugin(joiRequestValidatorPlugin)

        mockFn.mockReset()
      })

      test('coerces w/o option debug/warnings', async () => {
        app.post(
          '/ping',
          validate({
            body: Joi.object({
              count: Joi.number().integer().required(),
            }),
          }),
          (req, _res, next) => {
            mockFn(req.body)
            next()
          },
          handler
        )

        app.use(errorHandler)

        const response = await request(app).post('/ping').send({ count: '42' })
        expect(response.status).toBe(200)
        expect(mockFn.mock.calls).toMatchInlineSnapshot(`
          Array [
            Array [
              Object {
                "count": 42,
              },
            ],
          ]
        `)
      })

      test('coerces w/ option debug=true', async () => {
        app.post(
          '/ping',
          validate(
            {
              body: Joi.object({
                count: Joi.number().integer().required(),
              }),
            },
            { debug: true }
          ),
          (req, _res, next) => {
            mockFn(req.body)
            next()
          },
          handler
        )

        app.use(errorHandler)

        const response = await request(app).post('/ping').send({ count: '42' })
        expect(response.status).toBe(200)
        expect(mockFn.mock.calls).toMatchInlineSnapshot(`
          Array [
            Array [
              Object {
                "count": 42,
              },
            ],
          ]
        `)
      })

      test('coerces w/ option warnings=true', async () => {
        app.post(
          '/ping',
          validate(
            {
              body: Joi.object({
                count: Joi.number().integer().required(),
              }),
            },
            { warnings: true }
          ),
          (req, _res, next) => {
            mockFn(req.body)
            next()
          },
          handler
        )

        app.use(errorHandler)

        const response = await request(app).post('/ping').send({ count: '42' })
        expect(response.status).toBe(200)
        expect(mockFn.mock.calls).toMatchInlineSnapshot(`
          Array [
            Array [
              Object {
                "count": 42,
              },
            ],
          ]
        `)
      })

      test('does not coerce w/ option convert=false', async () => {
        app.post(
          '/ping',
          validate(
            {
              body: Joi.object({
                count: Joi.number().integer().required(),
              }),
            },
            { convert: false }
          ),
          (req, _res, next) => {
            mockFn(req.body)
            next()
          },
          handler
        )

        app.use(errorHandler)

        const response = await request(app).post('/ping').send({ count: '42' })
        expect(response.status).toBe(400)
        expect(response.body).toMatchSnapshot()
        expect(mockFn.mock.calls).toMatchInlineSnapshot(`Array []`)
      })
    })

    describe('options: segmentOrder', () => {
      test('(default) validates query before body', async () => {
        const expressOpenApi = new ExpressOpenAPI()

        app = express()
        app.use(express.json())

        const joiRequestValidatorPlugin = getJoiRequestValidatorPlugin({})

        validate = expressOpenApi.registerPlugin(joiRequestValidatorPlugin)

        app.post(
          '/ping',
          validate({
            body: {
              type: Joi.string()
                .valid('plastic', 'rubber', 'wood')
                .default('rubber'),
            },
            query: Joi.object({
              count: Joi.number().integer().required(),
            }),
          }),
          handler
        )

        app.use(errorHandler)

        const invalidResponse1 = await request(app)
          .post('/ping')
          .send({ type: 'paper' })
        expect(invalidResponse1.status).toBe(400)
        expect(invalidResponse1.body).toMatchSnapshot()

        const invalidResponse2 = await request(app)
          .post('/ping?count=42')
          .send({ type: 'paper' })
        expect(invalidResponse2.status).toBe(400)
        expect(invalidResponse2.body).toMatchSnapshot()

        const validResponse1 = await request(app)
          .post('/ping?count=42')
          .send({ type: 'wood' })
        expect(validResponse1.status).toBe(200)
      })

      test('(custom) validates body before query', async () => {
        const expressOpenApi = new ExpressOpenAPI()

        app = express()
        app.use(express.json())

        const joiRequestValidatorPlugin = getJoiRequestValidatorPlugin({
          segmentOrder: ['body', 'query'],
        })

        validate = expressOpenApi.registerPlugin(joiRequestValidatorPlugin)

        app.post(
          '/ping',
          validate({
            body: {
              type: Joi.string()
                .valid('plastic', 'rubber', 'wood')
                .default('rubber'),
            },
            query: Joi.object({
              count: Joi.number().integer().required(),
            }),
          }),
          handler
        )

        app.use(errorHandler)

        const invalidResponse1 = await request(app)
          .post('/ping')
          .send({ type: 'paper' })
        expect(invalidResponse1.status).toBe(400)
        expect(invalidResponse1.body).toMatchSnapshot()

        const invalidResponse2 = await request(app)
          .post('/ping')
          .send({ type: 'wood' })
        expect(invalidResponse2.status).toBe(400)
        expect(invalidResponse2.body).toMatchSnapshot()

        const validResponse1 = await request(app)
          .post('/ping?count=42')
          .send({ type: 'wood' })
        expect(validResponse1.status).toBe(200)
      })
    })
  })
})
