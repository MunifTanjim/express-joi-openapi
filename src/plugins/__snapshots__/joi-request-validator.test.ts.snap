// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getJoiRequestValidatorPlugin joiRequestValidatorPlugin options: segmentOrder (custom) validates body before query 1`] = `
Object {
  "error": Object {
    "details": Array [
      Object {
        "message": "\\"type\\" must be one of [plastic, rubber, wood]",
        "path": Array [
          "type",
        ],
        "type": "any.only",
      },
    ],
    "message": "\\"type\\" must be one of [plastic, rubber, wood]",
    "segment": "body",
  },
}
`;

exports[`getJoiRequestValidatorPlugin joiRequestValidatorPlugin options: segmentOrder (custom) validates body before query 2`] = `
Object {
  "error": Object {
    "details": Array [
      Object {
        "message": "\\"count\\" is required",
        "path": Array [
          "count",
        ],
        "type": "any.required",
      },
    ],
    "message": "\\"count\\" is required",
    "segment": "query",
  },
}
`;

exports[`getJoiRequestValidatorPlugin joiRequestValidatorPlugin options: segmentOrder (default) validates query before body 1`] = `
Object {
  "error": Object {
    "details": Array [
      Object {
        "message": "\\"count\\" is required",
        "path": Array [
          "count",
        ],
        "type": "any.required",
      },
    ],
    "message": "\\"count\\" is required",
    "segment": "query",
  },
}
`;

exports[`getJoiRequestValidatorPlugin joiRequestValidatorPlugin options: segmentOrder (default) validates query before body 2`] = `
Object {
  "error": Object {
    "details": Array [
      Object {
        "message": "\\"type\\" must be one of [plastic, rubber, wood]",
        "path": Array [
          "type",
        ],
        "type": "any.only",
      },
    ],
    "message": "\\"type\\" must be one of [plastic, rubber, wood]",
    "segment": "body",
  },
}
`;

exports[`getJoiRequestValidatorPlugin joiRequestValidatorPlugin segment validation validates body 1`] = `
Object {
  "error": Object {
    "details": Array [
      Object {
        "message": "\\"type\\" must be one of [plastic, rubber, wood]",
        "path": Array [
          "type",
        ],
        "type": "any.only",
      },
    ],
    "message": "\\"type\\" must be one of [plastic, rubber, wood]",
    "segment": "body",
  },
}
`;

exports[`getJoiRequestValidatorPlugin joiRequestValidatorPlugin segment validation validates headers 1`] = `
Object {
  "error": Object {
    "details": Array [
      Object {
        "message": "\\"authorization\\" with value \\"JWT 42\\" fails to match the required pattern: /^Bearer .+$/",
        "path": Array [
          "authorization",
        ],
        "type": "string.pattern.base",
      },
    ],
    "message": "\\"authorization\\" with value \\"JWT 42\\" fails to match the required pattern: /^Bearer .+$/",
    "segment": "headers",
  },
}
`;

exports[`getJoiRequestValidatorPlugin joiRequestValidatorPlugin segment validation validates params 1`] = `
Object {
  "error": Object {
    "details": Array [
      Object {
        "message": "\\"id\\" must be a number",
        "path": Array [
          "id",
        ],
        "type": "number.base",
      },
    ],
    "message": "\\"id\\" must be a number",
    "segment": "params",
  },
}
`;

exports[`getJoiRequestValidatorPlugin joiRequestValidatorPlugin segment validation validates query 1`] = `
Object {
  "error": Object {
    "details": Array [
      Object {
        "message": "\\"count\\" is required",
        "path": Array [
          "count",
        ],
        "type": "any.required",
      },
    ],
    "message": "\\"count\\" is required",
    "segment": "query",
  },
}
`;

exports[`getJoiRequestValidatorPlugin joiRequestValidatorPlugin segment validation validates query 2`] = `
Object {
  "error": Object {
    "details": Array [
      Object {
        "message": "\\"count\\" must be a number",
        "path": Array [
          "count",
        ],
        "type": "number.base",
      },
    ],
    "message": "\\"count\\" must be a number",
    "segment": "query",
  },
}
`;
